sky/provision/paperspace/instance.py

```diff
@@ -0,0 +1,297 @@
+    return common.ProvisionRecord(
+        provider_name='paperspace',
+        cluster_name=cluster_name_on_cloud,
+        region=region,
+        zone=None,
+        head_instance_id=head_instance_id,
+        resumed_instance_ids=list(stopped_instances.keys()),
+        created_instance_ids=created_instance_ids,
+    )
+
+
+def wait_instances(region: str, cluster_name_on_cloud: str,
+                   state: Optional[status_lib.ClusterStatus]) -> None:
+    del region, cluster_name_on_cloud, state  # unused
+    # We already wait on ready state in `run_instances` no need
+
+
+def stop_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    del provider_config, worker_only  # unused
+    client = utils.PaperspaceCloudClient()
+    all_instances = _filter_instances(cluster_name_on_cloud, [
+        'ready', 'serviceready', 'upgrading', 'provisioning', 'starting',
+        'restarting'
+    ])
+    num_instances = len(all_instances)
+
+    # Request a stop on all instances
+    for instance_id in all_instances:
+        client.stop(instance_id=instance_id)
+
+    # Wait for instances to stop
+    for _ in range(MAX_POLLS_FOR_STOP):
+        all_instances = _filter_instances(cluster_name_on_cloud, ['off'])
+        if len(all_instances) < num_instances:
+            break
+        time.sleep(POLL_INTERVAL)
+    else:
+        raise RuntimeError(f'Maximum number of polls: '
+                           f'{MAX_POLLS_FOR_STOP} reached. '
+                           f'Instance {all_instances} is still not in '
+                           'STOPPED status.')
+
+
+def terminate_instances(
+    cluster_name_on_cloud: str,
+    provider_config: Optional[Dict[str, Any]] = None,
+    worker_only: bool = False,
+) -> None:
+    """See sky/provision/__init__.py"""
+    del provider_config  # unused
+    client = utils.PaperspaceCloudClient()
+    instances = _filter_instances(cluster_name_on_cloud, None)
+    for inst_id, inst in instances.items():
+        logger.info(f'Terminating instance {inst_id}')
+        if worker_only and inst['name'].endswith('-head'):
+            continue
+        try:
+            client.remove(inst_id)
+        except Exception as e:  # pylint: disable=broad-except
+            with ux_utils.print_exception_no_traceback():
+                raise RuntimeError(
+                    f'Failed to terminate instance {inst_id}: '
+                    f'{common_utils.format_exception(e, use_bracket=False)}'
+                ) from e
+
+    if not worker_only:
+        try:
+            time.sleep(POLL_INTERVAL)
+            network = client.get_network(network_name=cluster_name_on_cloud)
+            client.delete_network(network_id=network['id'])
+        except IndexError:
+            logger.warning(f'Network {cluster_name_on_cloud}'
+                           'already deleted')
```

Comment:
This may not work if we are autodowning a cluster, where the head node will try to remove itself, and L217 will terminate itself first before it gets here. It might be fine for now, but we should add a TODO here for the leakage of the network.

Reply:
Would it be sufficient to just move the network cleanup logic before the instance teardown for-loop? That way there's no possibility of the head node removal being requested before the network is.